solutions/gold/cf-1181D.mdx

```diff
@@ -0,0 +1,101 @@
+---
+id: cf-1181D
+source: CF
+title: Irrigation
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/67727)
+
+## Explanation
+
+We can subtract $n$ from each of the queries, and answer them offline. Then, sort the cities by occurrences and queries by year.
+
+Let city $i$ denote the $i$'th smallest city sorted by occurrences, and $occ_i$ denote the number of occurrences. We can loop  through each $i$ and track the number of years has gone by. Assume cities $1, 2, ...i$ already have $occ_i$ occurrences. For them to all end up with $occ_{i+1}$ occurrences, we need $i \cdot (occ_{i+1} - occ_i)$ more years. This means that for the next $i \cdot (occ_{i+1} - occ_i)$ years, cities $1, 2, ... i$ will be selected one after another.
+
+How do we answer individual years? We can use a BST to track the initial indicies of each city $1, 2, 3... i$. Note that tie for occurrences is broken by initial indices, not the sorted.
+
+Let the current year be $y$, then at each iteration we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indicies. Since we have been tracking initial indicies with our BST, we can simply perform an index query for this information.
```

Comment:
```suggestion
Let the current year be $y$. At each iteration, we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indices. 
```